### PR TITLE
Update for assemble_emwave

### PR DESCRIPTION
**Pull request by acochrane**
revisions in-line with the 18-step recipe for new equations
Convert to complex permittivity
More inclusive intermask is necessary for Dirichlet BCs.
boundary condition for far-field direct-incidence excitation
boundary condition values come from input deck
bc for parallel
This update is to print logs to stderr if logging disabled at compile time
convert for phasors with exp(i[k_dot_r - omega t])
stabilization terms
robin condition
numerical jacobian and div squared stabilization
Sommerfeld Radiation Condition
Papanastasio 'Free' BC for EM equations
clear-up compile warnings

### CODE REVIEW

File: include/mm_names.h [hunk 1]
```diff
@@ -1594,6 +1605,18 @@ struct Equation_Names Exo_Var_Names[] =
...
+  { "Electric Field, Real z-component", "E3R", EM_E1_REAL },
+  { "Electric Field, Imaginary x-component", "E1I", EM_E1_IMAG },
+  { "Electric Field, Imaginary y-component", "E2I", EM_E1_IMAG },
+  { "Electric Field, Imaginary z-component", "E3I", EM_E1_IMAG },
+  { "Magnetic Field, Real x-component", "H1R", EM_H1_REAL },
+  { "Magnetic Field, Real y-component", "H2R", EM_H1_REAL },
+  { "Magnetic Field, Real z-component", "H3R", EM_H1_REAL },
+  { "Magnetic Field, Imaginary x-component", "H1I", EM_H1_IMAG },
+  { "Magnetic Field, Imaginary y-component", "H2I", EM_H1_IMAG },
+  { "Magnetic Field, Imaginary z-component", "H3I", EM_H1_IMAG },
```
Comment (wortiz) [inline include/mm_names.h]: I think changing these to ERX, ERY, ERZ, ... etc. you would get vector representation in paraview for free, might be worthwhile for visualization in the future

File: include/mm_fill_em.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+#ifdef GOMA_MM_FILL_EM_C
+#define EXTERN /* do nothing */
+#endif
+
+#ifndef GOMA_MM_FILL_EM_C
+#define EXTERN extern
+#endif
+
+#include <complex.h>
```
Comment (wortiz) [inline include/mm_fill_em.h]: Usually only include headers if they are used in the header would prefer moving to source file.

File: include/mm_fill_em.h [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  double [DIM][MAX_VARIABLE_TYPES+MAX_CONC][MDE] , // d_func
+  double [DIM] ,   // xi
+  const int);// bc_name
+
+
+EXTERN void calc_emwave_stabilization_term
+(struct emwave_stabilization*,
+ double);
+
+#undef I
```
Comment (wortiz) [inline include/mm_fill_em.h]: Would prefer this undef to be moved to source file (if it is still necessary) I couldn't find any other usage of I other than the int I indicated global node number we use sometimes, was that why this was required? Or was it something else?

File: src/mm_sol_nonlinear.c [hunk 4]
```diff
@@ -943,9 +943,10 @@ EH(-1,"version not compiled with frontal solver");
 				     &time_value, exo, dpi,
 				     &num_total_nodes,
 				     &h_elem_avg, &U_norm, NULL);
-	     
-              if( (vn->evssModel == LOG_CONF || vn->evssModel == LOG_CONF_GRADV)
-		  && pd->v[POLYMER_STRESS11] && af->Assemble_Jacobian == TRUE)
+
+              if( ( (vn->evssModel == LOG_CONF || vn->evssModel == LOG_CONF_GRADV)
+                 && pd->v[POLYMER_STRESS11] && af->Assemble_Jacobian == TRUE)
+                  || ( pd->v[EM_E1_REAL] && af->Assemble_Jacobian == TRUE) )
```
Comment (wortiz) [inline src/mm_sol_nonlinear.c]: Do you want it to be numerical Jacobians?

File: include/mm_fill_em.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+#define EXTERN /* do nothing */
+#endif
+
+#ifndef GOMA_MM_FILL_EM_C
+#define EXTERN extern
+#endif
+
+#include <complex.h>
+
+enum stabilization_type{none, phi_div, dphi_div, divphi_div, phi_divsquared, dphi_divsquared};
```
Comment (wortiz) [inline include/mm_fill_em.h]: enums add global constants would prefer if naming indicates this, something like

    enum em_stabilization_type{EM_NONE, EM_PHI_DIV, EM_DPHI_DIV, EM_DIVPHI_DIV, EM_PHI_DIVSQUARED, EM_DPHI_DIVSQUARED};

File: include/mm_fill_em.h (hunk 2 above)
Comment (acochrane) [reply include/mm_fill_em.h]: Ok, this and the #undef I was left-over from when I wanted to pass complex variables into functions.

File: src/mm_sol_nonlinear.c (hunk 4 above)
Comment (acochrane) [reply src/mm_sol_nonlinear.c]: For the time being, I think, yes. The Sommerfeld and free BC's don't have analytical Jacobian entries. With numerical Jacobians on, these features can be tested without any additional work/documentation. One day, when this set of equations is shown to work for some application, then the analytical Jacobians should be put in, and this line 949 will be removed.

File: include/mm_names.h (hunk 1 above)
Comment (wortiz) [reply include/mm_names.h]: I saw you changed this for Exo_Var_Names, I think you should also change it for Var_Names to match as I believe that's what we use for inputs like GD BCs and table BC and ACs so you probably want them to be the same

A lot of the time the second name in EQ_Name is the same as well

I feel like these needs to be condensed but that's for another time

**Comment by wortiz** [comment]
I think these need to be changed for numerical jacobian to actually work

This to include EM vars
https://github.com/acochrane/goma/blob/emwave2/src/mm_numjac.c#L585-L589

**Comment by acochrane** [comment]
> I think these need to be changed for numerical jacobian to actually work
> 
> This to include EM vars
> https://github.com/acochrane/goma/blob/emwave2/src/mm_numjac.c#L585-L589

It appears that you are correct.

Good News - Bad News.

When I make sure that line 598 gets hit to set the nj, I DO get convergence with my integrated boundary conditions that don't yet have analytic Jacobians.

But I get zero-row errors for my fixbc input deck. So I can't re-run the stabilization term comparison.

It makes me wonder, without setting nj, where did the updates come from when I ran my comparison of stabilization terms?

**Comment by rrrao** [comment]
Andrew,

I am not sure what conversation you are in the middle of, but it seems like you could get convergence without a good Jacobian as long as the entire row isn’t zero and the residuals started off small. Did you have some of the terms coded analytically?

Rekha

**Comment by wortiz** [comment]
> It makes me wonder, without setting nj, where did the updates come from when I ran my comparison of stabilization terms?

I mean you have some Jacobian terms from your assembly function, just missing stabilization term so it wasn't using that, if you converged to a good enough tolerance you probably wouldn't see any difference with the numerical Jacobian.

But that Jacobian was only good enough for a few cases hence lack of convergence for other problems. So it was just lucky that your Jacobian was good enough for some cases.

Fix BC sets the residual to 0 so we can't compute a numerical Jacobian, you could try setting BC_relax to something other than -1 for FIX_BC and see if that works, I think it would.

**Comment by rrrao** [comment]
I was under the impression the numerical one over-wrote the analytical.But if nj wasnt getting set, I suppose it might not have over-written the update. This is part of Goma I dont understand fully.Andrew CochraneAssistant Research ProfessorCOSMIAC at UNMAndrew.Cochrane@cosmiac.orgcell: (505) 876-7420Office: 1-844-COSMIAC2350 Alamo Avenue SE, Ste. 300Albuquerque, NM 87106Sent from my mobile device
-------- Original message --------From: "Rao, Rekha R" <rrrao@sandia.gov> Date: 3/10/20  3:28 PM  (GMT-07:00) To: goma/goma <reply@reply.github.com>, goma/goma <goma@noreply.github.com> Cc: Subscribed <subscribed@noreply.github.com>, Andrew Cochrane <andrew.cochrane@cosmiac.org> Subject: RE: [EXTERNAL] Re: [goma/goma] Update for assemble_emwave (#276) 

Andrew,
 
I am not sure what conversation you are in the middle of, but it seems like you could get convergence without a good Jacobian as long as the entire row isn’t zero and the residuals started off small. Did you have some of the terms coded
 analytically?
 
Rekha

**Comment by rrrao** [comment]
It sounds promising to add a look for EM variables similar to the stress look in the numerical Jacobian and see if it helps with Weston’s fix_BC idea.

From: andrew.cochrane <andrew.cochrane@cosmiac.org>
Sent: Tuesday, March 10, 2020 3:48 PM
To: Rao, Rekha R <rrrao@sandia.gov>; goma/goma <reply@reply.github.com>; goma/goma <goma@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: RE: [EXTERNAL] Re: [goma/goma] Update for assemble_emwave (#276)

I was under the impression the numerical one over-wrote the analytical.

But if nj wasnt getting set, I suppose it might not have over-written the update. This is part of Goma I dont understand fully.

Andrew Cochrane
Assistant Research Professor
COSMIAC at UNM
Andrew.Cochrane@cosmiac.org<mailto:Andrew.Cochrane@cosmiac.org>
cell: (505) 876-7420
Office: 1-844-COSMIAC

2350 Alamo Avenue SE, Ste. 300
Albuquerque, NM 87106

Sent from my mobile device

-------- Original message --------
From: "Rao, Rekha R" <rrrao@sandia.gov<mailto:rrrao@sandia.gov>>
Date: 3/10/20 3:28 PM (GMT-07:00)
To: goma/goma <reply@reply.github.com<mailto:reply@reply.github.com>>, goma/goma <goma@noreply.github.com<mailto:goma@noreply.github.com>>
Cc: Subscribed <subscribed@noreply.github.com<mailto:subscribed@noreply.github.com>>, Andrew Cochrane <andrew.cochrane@cosmiac.org<mailto:andrew.cochrane@cosmiac.org>>
Subject: RE: [EXTERNAL] Re: [goma/goma] Update for assemble_emwave (#276)

Andrew,

I am not sure what conversation you are in the middle of, but it seems like you could get convergence without a good Jacobian as long as the entire row isn’t zero and the residuals started off small. Did you have some of the terms coded analytically?

Rekha

**Comment by acochrane** [comment]
> > It makes me wonder, without setting nj, where did the updates come from when I ran my comparison of stabilization terms?
> 
> I mean you have some Jacobian terms from your assembly function, just missing stabilization term so it wasn't using that, if you converged to a good enough tolerance you probably wouldn't see any difference with the numerical Jacobian.

For some reason I was under the impression that the numerical Jacobian over-wrote the analytical one.

> But that Jacobian was only good enough for a few cases hence lack of convergence for other problems. So it was just lucky that your Jacobian was good enough for some cases.
> 
> Fix BC sets the residual to 0 so we can't compute a numerical Jacobian, you could try setting BC_relax to something other than -1 for FIX_BC and see if that works, I think it would.

I will investigate this, Thanks!

**Comment by wortiz** [comment]
Closing, should be partially taken care of by #311

**Pull request closed without merging**